tests/ui/malformed/do-not-ice-on-note_and_explain.rs

```diff
@@ -0,0 +1,7 @@
+struct A<B>(B);
+impl<B>A<B>{fn d(){fn d(){Self(1)}}}
```

Comment:
I feel like this is not the right way of fixing this bug? 

Why do we allow `Self` constructors from outer items that reference parameters that aren't in scope anyways? Like, we're referencing a type parameter that doesn't exist here, right?

I feel like this also may cause strange behavior (or unsoundness? not exactly sure, depends on when we compare params by name or by index) when we actually *do* have a substitutable param here:

```
use std::marker::PhantomData;

struct S0<T>(PhantomData<T>);

impl<T> S0<T> {
    fn foo() {
        fn bar<2T>() -> S0<T2> {
            Self(PhantomData) // <~ this constructs S0<T> not S0<T2> :(
        }
    }
}
```

cc @cjgillot who tried this and had to unland it in #111557

Reply:
https://github.com/rust-lang/rust/pull/111020 was reverted due to a regression, but it was probably a right thing to do in general.

Maybe we should do the usual bugfix process with a crater run, deprecation lint, and a tracking issue here.
In the meantime this PR could be merged to avoid the ICE.